### PR TITLE
feat: raft/log_test se testify packages in tests

### PR DESCRIPTION
**Pull request by falser101**
Signed-off-by: jianfei.zhang <jianfei.zhang@daocloud.io>

### CODE REVIEW

**Comment by falser101** [comment]
@fuweid
@ahrtr
help review, thanks!

File: raft/log_test.go [hunk 1]
```diff
@@ -47,11 +48,9 @@ func TestFindConflict(t *testing.T) {
...
 		raftLog.append(previousEnts...)
-
-		gconflict := raftLog.findConflict(tt.ents)
-		if gconflict != tt.wconflict {
-			t.Errorf("#%d: conflict = %d, want %d", i, gconflict, tt.wconflict)
-		}
+		t.Run(fmt.Sprint(i), func(t *testing.T) {
+			require.Equal(t, tt.wconflict, raftLog.findConflict(tt.ents))
+		})
 	}
```
Comment (ahrtr) [inline raft/log_test.go:54]: Please get all the code inside the `for` loop inside a subtest.

File: raft/log_test.go [hunk 2]
```diff
@@ -79,10 +78,9 @@ func TestIsUpToDate(t *testing.T) {
...
 
 	for i, tt := range tests {
-		gUpToDate := raftLog.isUpToDate(tt.lastIndex, tt.term)
-		if gUpToDate != tt.wUpToDate {
-			t.Errorf("#%d: uptodate = %v, want %v", i, gUpToDate, tt.wUpToDate)
-		}
+		t.Run(fmt.Sprint(i), func(t *testing.T) {
+			require.Equal(t, tt.wUpToDate, raftLog.isUpToDate(tt.lastIndex, tt.term))
+		})
```
Comment (ahrtr) [inline raft/log_test.go:83]: ditto

File: raft/log_test.go [hunk 3]
```diff
@@ -127,20 +125,13 @@ func TestAppend(t *testing.T) {
...
-		if goff := raftLog.unstable.offset; goff != tt.wunstable {
-			t.Errorf("#%d: unstable = %d, want %d", i, goff, tt.wunstable)
-		}
+		t.Run(fmt.Sprint(i), func(t *testing.T) {
+			require.Equal(t, tt.windex, raftLog.append(tt.ents...))
+			g, err := raftLog.entries(1, noLimit)
+			require.NoError(t, err)
+			require.Equal(t, tt.wents, g)
+			require.Equal(t, tt.wunstable, raftLog.unstable.offset)
+		})
```
Comment (ahrtr) [inline raft/log_test.go:134]: ditto

**Comment by falser101** [comment]
done!@ahrtr

**Review by ahrtr — APPROVED** [review]
LGTM

Thank you @falser101

File: raft/log_test.go [hunk 4]
```diff
@@ -546,36 +499,33 @@ func TestCompaction(t *testing.T) {
...
-
-			for j := 0; j < len(tt.compact); j++ {
-				err := storage.Compact(tt.compact[j])
-				if err != nil {
-					if tt.wallow {
-						t.Errorf("#%d.%d allow = %t, want %t", i, j, false, tt.wallow)
+			if !tt.wallow {
+				require.Panics(t, func() {
+					for j := 0; j < len(tt.compact); j++ {
+						_ = storage.Compact(tt.compact[j])
```
Comment (fuweid) [inline raft/log_test.go:507]: I think we should not change the test case when we refactor.
We should prepare the log before `Compact`.

It should be like this, if we have to use `require.Panics`.

```go
storage := NewMemoryStorage()
for i := uint64(1); i <= tt.lastIndex; i++ {
    storage.Append([]pb.Entry{{Index: i}})
}
raftLog := newLog(storage, raftLogger)
raftLog.maybeCommit(tt.lastIndex, 0)
raftLog.appliedTo(raftLog.committed)

verifyIn := require.NoPanics
if !tt.wallow {
   verifyIn = require.Panics
}

verifyIn(t, func() {
   for j := 0; j < len(tt.compact); j++ {
	err := storage.Compact(tt.compact[j])
	if err != nil {
		require.False(t, tt.wallow)
		continue
	}

	require.Equal(t, tt.wleft[j], len(raftLog.allEntries()))
    }
})
```

File: raft/log_test.go [hunk 5]
```diff
@@ -453,22 +410,17 @@ func TestCommitTo(t *testing.T) {
...
-			}()
-			raftLog := newLog(NewMemoryStorage(), raftLogger)
-			raftLog.append(previousEnts...)
-			raftLog.committed = commit
-			raftLog.commitTo(tt.commit)
-			if raftLog.committed != tt.wcommit {
-				t.Errorf("#%d: committed = %d, want %d", i, raftLog.committed, tt.wcommit)
+		t.Run(fmt.Sprint(i), func(t *testing.T) {
+			if tt.wpanic {
+				newLog(NewMemoryStorage(), raftLogger)
```
Comment (fuweid) [inline raft/log_test.go:415]: I don't think it is correct.

The panic happens when tt.commit > 2 [`commit := uint64(2)`]

File: raft/log_test.go (hunk 4 above)
Comment (fuweid) [reply raft/log_test.go:507]: Just my two cents, use `defer` looks more clear than using `Panics` 😂

Comment (falser101) [reply raft/log_test.go:507]: good ieda, let me resolve that

**Review by ahrtr — CHANGES_REQUESTED** [review]
Please fix the unit test failures.

**Comment by falser101** [comment]
@ahrtr @fuweid done!

**Comment by falser101** [comment]
@fuweid @ahrtr done！

**Review by fuweid — APPROVED** [review]
LGTM on green (non-binding)

**Review by ahrtr — APPROVED** [review]
Thanks both @falser101 and @fuweid

**Pull request merged**